### ISSUE TITLE
bug(compiler-cli): metadata resolution error when generating flat module with reference in re-export

### ISSUE THREAD

**Issue by trotyl** [angular/angular#20931]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

#### GIVEN

There is a `HelloComponent` in `app/components/hello.component`.
There is a re-export of `HelloComponent` in `app/components.ts`, like:

```typescript
export * from './components/hello.component'
```

In `app.module.ts`, import `HelloComponent` from `app/components.ts` (re-export), like:

```typescript
// ...
import { HelloComponent } from './components'

@NgModule({
  declarations: [
    // ...
    HelloComponent,
  ],
  // ...
})
export class AppModule { }
```

<sub>Checked not relates to filename.</sub>

#### WHEN

Config it to generate a flat module:

```json
{
  "angularCompilerOptions": {
    "flatModuleId": "hello",
    "flatModuleOutFile": "hello.js",
    "skipTemplateCodegen": true
  }
}
```

And call AOT compiler via `ngc`.

#### THEN

An error occur:

```text
MACHINE_NAME:aot-reexport-issue zjyu$ $(yarn bin)/ngc -p src/tsconfig.app.json
: TypeError: Cannot read property 'module' of undefined
    at MetadataBundler.convertSymbol (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:264:57)
    at createReference (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:390:23)
    at MetadataBundler.convertReference (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:438:24)
    at MetadataBundler.convertExpression (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:367:33)
    at MetadataBundler.convertValue (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:348:25)
    at /Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:351:58
    at Array.map (<anonymous>)
    at MetadataBundler.convertValue (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:351:26)
    at MetadataBundler.convertValue (/Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:357:32)
    at /Users/USER_NAME/Projects/github/trotyl/aot-reexport-issue/node_modules/@angular/compiler-cli/src/metadata/bundler.js:351:58
```

The error is caused by [this line](https://github.com/angular/angular/blob/a75040d0a15c06cc8250032f044b7df0ed885317/packages/compiler-cli/src/metadata/bundler.ts#L343), `canonicalSymbol` doesn't have a `declaration` property in this scenario. 

## Expected behavior

Be able to compile.

## Minimal reproduction of the problem with instructions

https://github.com/trotyl/aot-reexport-issue

Steps mentioned there.

## What is the motivation / use case for changing the behavior?

Be able to get codes work.

## Environment

<pre><code>
Angular version: 5.1.0

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.2  
- Platform:  Mac 

Others:

</code></pre>

**Comment by rohitpratapitm** [comment]
Is there any workaround ?

**Comment by leosvelperez** [comment]
Any updates?

**Comment by babeal** [comment]
We are running into this issue as well (ng-packagr uses this approach rendering the solution useless).  Just with a single SharedModule that is reexported through an index file. It breaks.  The error is thrown on line 344 of the bundler.ts file ```const module = this.getMetadata(declaration.module);``` . declaration is undefined.  The declaration is found on the symbol just not on the canonicalSymbol.

**Comment by kolesnik** [comment]
+1

**Comment by deepthan** [comment]
One reference at a time will succeed.
```
import a from 'AComponent';
import b from 'BComponent';
```

**Comment by TheAkio** [comment]
What @deepthan mentions seems to be the solution right now. I was trying to re-export decorator functions so i don't have to import them one by one but files re-exporting files don't work at all.

**Comment by jjanuszkiewicz** [comment]
It is a problem for me as well, with `@angular/cli` 7.2.6 / `@angular/cli` 7.3.3. Is there any plan to fix this?

**Comment by Tahiche** [comment]
> Importing it from where it is defined is work for me.

Hi, ¿can you explain more in detail?.

**Comment by xjose97x** [comment]
Any update here?

**Comment by alxpsr** [comment]
😢

**Comment by alxpsr** [comment]
I noticed unexpected (for me) behavior:

```
//== Class that use interface from barrel-file
import { IBodyInterface } from '../interfaces'; // => Import from barrel

export class MyModel {
  public Body: IBodyInterface;
}

//== Barrel
export { IBodyInterface } from './body.interface';
```

In this case i got error `TypeError: Cannot read property 'module' of undefined`.
But if edit export string in barrel file to `export * from './body.interface';` build error will disappear

**Comment by raphael22** [comment]
Any ETA on this ?
Resolving from index barrel is kind of important on large, multi-teams project...

**Comment by benwinding** [comment]
Instead of `index.ts` files, 
``` ts
import { SomeComponent, SomeService } from '../something';
```
we're using private _Angular Modules_
``` ts
import { SomeModule } from '../something.module';
```
This helps in a few cases to reduce _direct referencing_.

But `index.ts` files would be much better!!!!

**Comment by JoostK** [comment]
Marking as fixed by Ivy but still an issue today, as libraries are still built for VE therefor dependent on the metedata bundler.

**Comment by alxhub** [comment]
I don't have much to add beyond @JoostK - soon this will not be a problem for Ivy, but unfortunately it still is an issue today.

**Comment by ersimont** [comment]
I was getting this error in my angular library when I imported like:

```ts
import { thing } from 'directory';
```

But this workaround fixed it for me:

```ts
import { thing } from 'directory/index';
```

**Comment by kroeder** [comment]
This issue prevents my library to use secondary-entry points and would save a lot of package size because of non-tree-shakable third-party libs. Neither is Ivy a thing for library-authors yet nor is there a workaround for this issue (at least I could not get it working)

How can this be a `very low priority` issue since Angular is working on low bundle sizes since forever? Not meant in an offensive way.

# Addition to this issue

I tried to create two secondary entry points. One does work, one does not and I can't see a difference.

## This does work

### `ui/charts/package.json`

```json
{
  "ngPackage": {
    "lib": {
      "entryFile": "../src/lib/components/data/chart/public_api.ts",
      "umdModuleIds": {
        "chart.js": "chart.js"
      }
    }
  },
  "peerDependencies": {
    "chart.js": "^2.8.0"
  }
}
```

### `public_api.ts`

```typescript
export { ChartModule } from './nx-chart.module';
export { ChartBarComponent } from './chart-bar/chart-bar.component';
export { ChartLineComponent } from './chart-line/chart-line.component';
```

## This does not work

### `ui/object-editor/package.json`

```json
{
  "ngPackage": {
    "lib": {
      "entryFile": "../src/lib/components/inputs/object-editor/public_api.ts",
      "umdModuleIds": {
        "ang-jsoneditor": "ang-jsoneditor",
        "jsoneditor": "jsoneditor"
      }
    }
  },
  "peerDependencies": {
    "ang-jsoneditor": "^1.9.0",
    "jsoneditor": "^8.0.0"
  }
}
```

### `public_api.ts`

```typescript
export { ObjectEditorModule } from './object-editor.module';
export { ObjectEditorComponent } from './object-editor.component';
export { ObjectEditorSectionComponent } from './object-editor-section/object-editor-section.component';
```

## Conclusion

The only difference I spot is the dependencies of both entry points. I'd be happy to help with further information if you need them.

**Comment by kroeder** [comment]
I figured out where this error is thrown
https://github.com/angular/angular/blob/master/packages/compiler-cli/src/metadata/bundler.ts#L358

```typescript
// declaration is ensured to be definded before this method is called.
const declaration = canonicalSymbol.declaration!; // <------- declaration is not defined in my case
```

What ensures this? Without a hint to that location, I can't figure out why it is not defined
`declaration is ensured to be defined before this method is called.` 

@alxhub any idea? I'd be more than happy to help out here.


**canonicalSymbol** is in my case:

```javascript
{
  canonicalSymbol: {
    module: '../../overlays/dialog/dialog.module',
    name: 'DialogModule',
    reexportedAs: {
      module: '../../overlays/dialog/public_api',
      name: 'DialogModule',
      exports: [Circular],
      isPrivate: true,
      declaration: [Circular],
      canonicalSymbol: [Circular],
      reexport: undefined
    },
    referenced: true
  }
}
```

I can't see any odd things happening in `DialogModule`

```typescript
@NgModule({
    imports: [
        CommonModule,
        NextCommonModule,
        FooterModule,
        NextClientInjectorModule,
        OverlayModule
    ],
    exports: [OverlayDialogTitleComponent],
    declarations: [DialogComponent, OverlayDialogTitleComponent],
    entryComponents: [DialogComponent],
    providers: [DialogService]
})
export class DialogModule {
}
```

**Comment by kroeder** [comment]
I figured out that `canonicalSymbol.declaration` is in most cases undefined. I found only one other symbol that had `declaration` assigned. However this also means that it is not needed most of the time.

Here's a try-and-error workaround that at least gave me correct compiler errors 

```typescript
                if(!canonicalSymbol.declaration) {
                    canonicalSymbol.declaration = { ...symbol.declaration };
                }
                var declaration = canonicalSymbol.declaration;
```

Now I get lots of `needs to be under root-dir` issues of my dependencies. This actually makes a lot of sense now. 
My other chart package does not have any external dependencies of my library.

I don't know if the `module undefined` error is related to the `needs to be under root-dir` error. Maybe the rootDir check needs to be checked earlier in order to prevent the `module of undefined` error? Just a theory and most likely not true. I tried to reproduce this error in a new project and got the `rootDir` errors without the `module undefined` error

**Comment by kroeder** [comment]
**Update**

`ERROR: Cannot read property 'module' of undefined` was actually caused by a 'bug' in my library.

After trying my above mentioned workaround (https://github.com/angular/angular/issues/20931#issuecomment-756782380) I then got import errors `needs to be under root-dir` 

A long time ago we decided to not using package-name imports within a package `import { Foo } from 'foo-package';` but instead `import { Foo } from '../relative-path/to/component'` and black-list self-imports. I'm not sure why anymore.

However, if I change it back to `import { Foo } from 'foo-package'` and remove all relative paths I at least get correct errors and the `module of undefined` error is gone.

Hope this helps! Check your imports. 
I'm still curious if compiler-cli can catch this error properly before everything crashes 🤔
